pkg/typedef/typedef.go

```diff
@@ -197,32 +197,26 @@ const (
 	CacheArrayLen
 )
 
-func prettyCQL(query string, values Values, types Types) string {
-	if len(values) == 0 {
+func prettyCQL(query string, values Values, types []Type) string {
+	if len(types) == 0 {
 		return query
 	}
 
-	k := 0
-	out := make([]string, 0, len(values)*2)
-	queryChunks := strings.Split(query, "?")
-	out = append(out, queryChunks[0])
-	qID := 1
-	for _, typ := range types {
-		tupleType, ok := typ.(*TupleType)
-		if !ok {
-			out = append(out, typ.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
-			continue
-		}
-		for _, t := range tupleType.ValueTypes {
-			out = append(out, t.CQLPretty(values[k]))
-			out = append(out, queryChunks[qID])
-			qID++
-			k++
-		}
+	var (
+		index   int
+		builder strings.Builder
+	)
+
+	builder.Grow(len(query))
+
+	for pos, i := strings.Index(query[index:], "?"), 0; pos != -1; pos, i = strings.Index(query[index:], "?"), i+1 {
+		actualPos := index + pos
+		builder.WriteString(query[index:actualPos])
+		types[i].CQLPretty(&builder, values[i])
+
+		index = actualPos + 1
 	}
```

Comment:
Tuples implementation, unfortunately is totally different from other types.
For all the types there is a single value that is associated with this given type.
For tuples you have list of values, associated with it, every tuple field gets it's own value.

So, you can't just take one value for tuple, you need to take list of values and feed them to the `types[i].CQLPretty`

Reply:
To me, tuples should be like any other type, but enclosed in `( )` instead of the list of `?` One should be enough to do proper formatting, and it seems that this behavior is what Gemini already uses without any problems.

when we have some query like this
`SELECT * FROM tbl WHERE pk=?` is equivalent to `SELECT * FROM tbl WHERE pk=(?,?)` if we do the tuple formatting correctly, this removes the special case, and also tuples should be in `()` instead of `{}`, it looks like ScyllaDB does not make a distinction, and accepts both in `SELECT`

> So, you can't just take one value for tuple, you need to take list of values and feed them to the types[i].CQLPretty

Values for the tuples already are formatted as `[]any` inside `[]any` and not as separate values, but I'll do more testing to see if this is the case, to be 100% sure.

Reply:
> To me, tuples should be like any other type, but enclosed in `( )` instead of the list of `?` One should be enough to do proper formatting, and it seems that this behavior is what Gemini already uses without any problems.

Agree, should be, but it is not the case, gocql expects to see multiple values for a tuple type.

> 
> when we have some query like this `SELECT * FROM tbl WHERE pk=?` is equivalent to `SELECT * FROM tbl WHERE pk=(?,?)` if we do the tuple formatting correctly, this removes the special case, and also tuples should be in `()` instead of `{}`, it looks like ScyllaDB does not make a distinction, and accepts both in `SELECT`
> 
> > So, you can't just take one value for tuple, you need to take list of values and feed them to the types[i].CQLPretty
> 
> Values for the tuples already are formatted as `[]any` inside `[]any` and not as separate values, but I'll do more testing to see if this is the case, to be 100% sure.

It is quite oposite, take a look here: https://github.com/scylladb/gemini/blob/ffb469f5843673a23f39e5a340c3a38cf471550e/pkg/typedef/tuple.go#L87-L101